keras_cv/models/segmentation/deeplabv3plus.py

```diff
@@ -0,0 +1,354 @@
+                filters=256,
+                convs=2,
+                dropout=0.2,
+                kernel_size=3,
+            )
+
+        output = segmentation_head(output)
+
+        super().__init__(
+            inputs={
+                "inputs": inputs,
+            },
+            outputs={
+                "output": output,
+            },
+            **kwargs,
+        )
+
+        # All references to `self` below this line
+        self.classes = classes
+        self.backbone = backbone
+        self.spatial_pyramid_pooling = spatial_pyramid_pooling
+        self.segmentation_head = segmentation_head
+        self.segmentation_head_activation = segmentation_head_activation
+        self.layer_names = layer_names
+
+    def get_config(self):
+        return {
+            "classes": self.classes,
+            "backbone": self.backbone,
+            "spatial_pyramid_pooling": self.spatial_pyramid_pooling,
+            "segmentation_head": self.segmentation_head,
+            "segmentation_head_activation": self.segmentation_head_activation,
+            "layer_names": self.layer_names,
+        }
+
+    def compile(self, weight_decay=0.0001, **kwargs):
+        self.weight_decay = weight_decay
+        super().compile(**kwargs)
+
+    def train_step(self, data):
+        images, y_true, sample_weight = tf.keras.utils.unpack_x_y_sample_weight(data)
+        with tf.GradientTape() as tape:
+            y_pred = self(images, training=True)
+            total_loss = self.compute_loss(images, y_true, y_pred, sample_weight)
+            reg_losses = []
+            if self.weight_decay:
+                for var in self.trainable_variables:
+                    if "bn" not in var.name:
+                        reg_losses.append(self.weight_decay * tf.nn.l2_loss(var))
+                l2_loss = tf.math.add_n(reg_losses)
+                total_loss += l2_loss
+        self.optimizer.minimize(total_loss, self.trainable_variables, tape=tape)
+        return self.compute_metrics(images, y_true, y_pred, sample_weight=sample_weight)
+
+
+class SegmentationHead(layers.Layer):
+    """Prediction head for the segmentation model
+
+    The head will take the output from decoder (eg FPN or ASPP), and produce a
+    segmentation mask (pixel level classifications) as the output for the model.
+
+    Args:
+        classes: int, the number of output classes for the prediction. This should
+            include all the classes (eg background) for the model to predict.
+        convs: int, the number of conv2D layers that are stacked before the final
+            classification layer. Default to 2.
+        filters: int, the number of filter/channels for the the conv2D layers. Default
+            to 256.
+        activations: str or 'tf.keras.activations', activation functions between the
+            conv2D layers and the final classification layer. Default to 'relu'
+        output_scale_factor: int, or a pair of ints, for upsample the output mask.
```

Comment:
output mask or output/prediction?

Reply:
Output prediction, but the prediction itself is effectively a mask.
This was left from the original SegHead docstrings by Scott

Reply:
Output prediction, but the prediction itself is effectively a mask.
This was left from the original SegHead docstrings by Scott